domain/application/state/containerimagemetadata.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+
+	"github.com/canonical/sqlair"
+
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/logger"
+	coreresource "github.com/juju/juju/core/resources"
+	"github.com/juju/juju/core/resources/containermetadataresource"
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/domain/application"
+	applicationerrors "github.com/juju/juju/domain/application/errors"
+	"github.com/juju/juju/internal/errors"
+)
+
+// ContainerImageMetadataState describes retrieval and persistence methods for storage.
+type ContainerImageMetadataState struct {
+	*domain.StateBase
+	logger logger.Logger
+}
+
+// NewContainerImageMetadataState returns a new state reference.
+func NewContainerImageMetadataState(factory database.TxnRunnerFactory, logger logger.Logger) *ContainerImageMetadataState {
+	return &ContainerImageMetadataState{
+		StateBase: domain.NewStateBase(factory),
+		logger:    logger,
+	}
+}
+
+// Get returns the container image metadata with the given UUID.
+// applicationerrors.ResourceNotFound is returned if the UUID is not in the
+// table
+func (s *ContainerImageMetadataState) Get(
+	ctx context.Context,
+	resourceUUID coreresource.ID,
+) (application.ContainerImageMetadata, error) {
+	db, err := s.DB()
+	if err != nil {
+		return application.ContainerImageMetadata{}, errors.Capture(err)
+	}
+
+	storageKey := containerImageMetadataStorageKey{StorageKey: resourceUUID.String()}
+	m := containerImageMetadata{}
+	stmt, err := s.Prepare(`
+SELECT &containerImageMetadata.*
+FROM   resource_oci_image_metadata_store
+WHERE  storage_key = $containerImageMetadataStorageKey.storage_key
+`, storageKey, m)
+	if err != nil {
+		return application.ContainerImageMetadata{}, errors.Errorf(
+			"preparing select OCI image resource metadata statement: %w", err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, storageKey).Get(&m)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return applicationerrors.ResourceNotFound
```

Comment:
**question:** I am ok with that but I wonder if we shouldn't have a OCIImageNotFound error, or some store-related error ?

Reply:
Good point, I've added that. In the service I want to capture this error and return a generic "stored resource not found" that is good for both the object store and this.